### PR TITLE
Split compute budget instructions process from struct

### PR DESCRIPTION
**Pull request by taozhu-chicago**
#### Problem

#33513 had bank hash mismatch on canary (https://github.com/solana-labs/solana/pull/33513#issuecomment-1772733600). Change was reverted by #33784 

#### The cause: 
TX uses deprecated instruction to request compute units greater than Max 1.4M CU, examples below, resulting compute_unit_price calculated lower than MNB version (compute_unit_price = additional_fee / cu_limit). This in turn caused priority fee were charged differently between branch and MNB.

example: https://solscan.io/tx/x1Fi6X3FmYUaqpS4DRsk5kR4RjK6jW5dHyyHRMHNLUCSdwwxmxz8RgfrqyYEZRKc5Kwxbxq2TgPYGWrwygAA6qW
https://solscan.io/tx/5usU3X3LRMywsxEkUW15HFXZ5PgAYWBnsiUGU8JRFyyvDjycxzCEeRQ1oWT5bYk14vfmjAzAtpHJzpixZNcusfxJ

#### Summary of Changes
All commits are reviewed changes, except the last one, 5104d08c25dc83897c2f795c1f4a0642e334add9, that fixes the mismatched hash. It does: 
- Add test to illustrate the corner cases
- Add optional `deprecated_additional_fee` to `ComputeBudgetLimits` to support deprecated instruction 

#### ledger-tool replay result: 
Thanks @steviez for providing ledger for repro and confirm fix. 

```
[2023-10-24T23:05:22.149511371Z INFO  solana_runtime::bank] bank frozen: 224820287 hash: DwtKodrR4LxqZfbP6WtBrjteWQi9do9cSHPNAHEsnJc3 accounts_delta: 3mJ9VLXJgLFnE8iPF8W9CSLpfxvdC3NUfi6k1fWhjTnu signature_count: 363 last_blockhash: G2ZkKofPKbrobQ3X4H77cbmfPqYjwzvyWcv7rqUi3psW capitalization: 560381532334278680, stats: BankHashStats { num_updated_accounts: 1897, num_removed_accounts: 8, num_lamports_stored: 126533203599805, total_data_len: 6248717, num_executable_accounts: 0 }
```

### CODE REVIEW

File: program-runtime/src/compute_budget_processor.rs [hunk 1]
```diff
@@ -0,0 +1,704 @@
...
+        deprecated_additional_fee,
+    })
+}
+
+fn sanitize_requested_heap_size(bytes: u32) -> bool {
+    (u32::try_from(MIN_HEAP_FRAME_BYTES).unwrap()..=MAX_HEAP_FRAME_BYTES).contains(&bytes)
+        && bytes % 1024 == 0
+}
+
+// Supports request_units_derpecated ix, returns compute_unit_price from deprecated requested
```
Comment (apfitzge) [inline program-runtime/src/compute_budget_processor.rs:191]:
```suggestion
// Supports request_units_deprecated ix, returns compute_unit_price from deprecated requested
```

File: program-runtime/src/compute_budget_processor.rs [hunk 2]
```diff
@@ -0,0 +1,704 @@
...
+        // a normal case
+        try_prioritization_fee_from_deprecated_requested_units(647, 6002);
+
+        // requesting cu limit more than MAX, div result will be round down
+        try_prioritization_fee_from_deprecated_requested_units(
+            640,
+            MAX_COMPUTE_UNIT_LIMIT + 606_002,
+        );
+
+        // requesting cu limit more than MAX, div resoult will round up
```
Comment (apfitzge) [inline program-runtime/src/compute_budget_processor.rs:698]:
```suggestion
        // requesting cu limit more than MAX, div result will round up
```

File: runtime/src/bank.rs [hunk 3]
```diff
@@ -5179,36 +5180,28 @@ impl Bank {
...
+                            saturating_add_assign!(
+                                timings
+                                    .execute_accessories
+                                    .compute_budget_process_transaction_us,
+                                compute_budget_process_transaction_time.as_us()
+                            );
+                            if let Err(err) = maybe_compute_budget {
+                                return TransactionExecutionResult::NotExecuted(err);
+                            }
+                            maybe_compute_budget.unwrap()
```
Comment (apfitzge) [inline runtime/src/bank.rs:5203]: nit: we're already inspecting before this `unwrap`, hence why it's safe to do; but it might be better as something like:

```rust
maybe_compute_budget.map_err(|err| TransactionExecutionResult::NotExecuted(err)?
```

**Review by apfitzge — DISMISSED** [review]
Just re-reviewed the new commit which fixes the hash mismatch.
I didn't previously realize/notice the difference in behavior for the deprecated request. A follow-up questions from the issue:

What's the reason we want to allow higher than 1.4M requests (rounded down) rather than just failing them?

File: program-runtime/src/compute_budget_processor.rs [hunk 4]
```diff
@@ -616,4 +627,78 @@ mod tests {
...
+            u64::from(additional_fee)
+        );
+        assert_eq!(
+            fee_budget_limits.compute_unit_limit,
+            u64::from(expected_compute_unit_limit)
+        );
+    }
+
+    #[test]
+    fn test_support_deprecated_requested_units() {
```
Comment (steviez) [inline program-runtime/src/compute_budget_processor.rs:688]: To confirm, does this unit test fail before the addition of `deprecated_additional_fee`? I don't think it is necessary to retroactively split the commits on this PR, but in general, it can be very nice to break things up into two+ commits:
- The first commit that introduces a test that fails with current state (typically master or in this PR's case, the commits from the initial PR)
- The subsequent commit(s) that address the issue (ie get the unit test passing)

Again, I don't think strictly necessary to split the commits for this PR, but maybe you can tweak things locally to run the unit test without the fix and confirm you see the failure as an extra sanity check. The fact that you see the correct bank hash on the mnb ledger seemingly indicates you have resolved things, but hopefully this is something you can run quickly that will give us an extra sanity check

**Comment by taozhu-chicago** [comment]
> What's the reason we want to allow higher than 1.4M requests (rounded down) rather than just failing them?

I don't know the reason, probably didn't think people would ask more than 1.4M ? In any case, can't change that behavior now without a feature gate. The good news is the feature gate to remove that deprecate instruction is almost there for mnb (already activated in testnet and devnet).

File: program-runtime/src/compute_budget_processor.rs (hunk 4 above)
Comment (taozhu-chicago) [reply program-runtime/src/compute_budget_processor.rs:688]: Yep, I added this test first to confirm the theory; Ran it and confirmed it failed with exact priority fee as I got from ledge-tool replay that resulted mismatching hash. Then I did the fix (adding optional `deprecated_additional_fee`) to green the test. 

I agree that separate commits are better. In this case, I thought appending one self-contained commit to already reviewed changes would reduce cluster, so I did `git rebase -i HEAD~2`. Look back, keeping 2 commits ain't that bad at all :)

Comment (steviez) [reply program-runtime/src/compute_budget_processor.rs:688]: Good deal / thanks for confirming!

**Review by steviez — APPROVED** [review]
This looks good to me from the standpoint of correcting the consensus divergence. Given that @apfitzge reviewed the original PR, I'm going to defer to him for the finer details of this PR. So, please get a ship it from Andrew before pushing this

**Review by apfitzge — APPROVED** [review]
Still lgtm
